### ISSUE TITLE
Mini-Batch vs Batch

### ISSUE THREAD

**Issue by SachinKonan** [NovaSky-AI/SkyRL#872]
hi, i have been setting these configs during my training:
```
trainer.train_batch_size=128 \
trainer.policy_mini_batch_size=64 \
```

my understanding is that we perform |data_size|/64 optimization steps. Right now, the len of the dataloader depends on the train_batch_size, not the mini-batch-size. So does this means we decay 2x as fast (because it happens after every optimization step).

**Comment by erictang000** [comment]
hey @SachinKonan 

> my understanding is that we perform |data_size|/64 optimization steps. 

yes that's right

> Right now, the len of the dataloader depends on the train_batch_size, not the mini-batch-size.

Yep the train dataloader: https://github.com/NovaSky-AI/SkyRL/blob/8809eea6cd49eb11daa43aaba840627396386949/skyrl-train/skyrl_train/trainer.py#L193 operates over the train batch size

> So does this means we decay 2x as fast (because it happens after every optimization step).

can you elaborate what you mean by this? do you mean in the case where we use a learning rate scheduler like cosine decay whether it operates over # of training batches or # of mini batches?

**Comment by SachinKonan** [comment]
> do you mean in the case where we use a learning rate scheduler like cosine decay whether it operates over # of training batches or # of mini batches?

Yup this is what im wondering. Based off my logs, I was thinking the latter (number of mini-batches), but wanted to confirm.

**Comment by erictang000** [comment]
yup you're right, it's # of mini batches 

scheduler.step() is called on each call to optimizer.step():

https://github.com/NovaSky-AI/SkyRL/blob/8809eea6cd49eb11daa43aaba840627396386949/skyrl-train/skyrl_train/distributed/fsdp_strategy.py#L190

num_warmup_steps for optimizer similarly works off of the # of mini batch steps
https://github.com/NovaSky-AI/SkyRL/blob/8809eea6cd49eb11daa43aaba840627396386949/skyrl-train/skyrl_train/config/ppo_base_config.yaml#L43

**Comment by SachinKonan** [comment]
cool thanks for confirming. so if i wanted to decay with respect to the number of optimization steps, then i could just this in my config: **num_training_steps** https://huggingface.co/docs/transformers/en/main_classes/optimizer_schedules#transformers.get_cosine_schedule_with_warmup

which would be size of my dataset/mini_batch_size ?


Edit: I dont think this is possible because num_training_steps isn't available in the `optimizer_config` of `ppo_base_config.yaml`. are there any other options?

**Comment by erictang000** [comment]
@SachinKonan hmm yeah good catch, it looks like there's a discrepancy between what's being passed down to the scheduler vs how it's getting `step()` called on it.

num_training_step gets calculated here: 
https://github.com/NovaSky-AI/SkyRL/blob/8809eea6cd49eb11daa43aaba840627396386949/skyrl-train/skyrl_train/trainer.py#L120

based on the number of training batches (and propagated down here: https://github.com/NovaSky-AI/SkyRL/blob/8809eea6cd49eb11daa43aaba840627396386949/skyrl-train/skyrl_train/trainer.py#L507)

but stepped based on the number of mini batch steps.

I think we can just fix the value that gets propagated down by multiplying the number of total training steps by the number of mini batches per training step.

**Comment by SachinKonan** [comment]
yeah i think that makes sense. i think right now, because of this, there is a discrepancy with the async trainer:

https://github.com/NovaSky-AI/SkyRL/blob/8809eea6cd49eb11daa43aaba840627396386949/skyrl-train/skyrl_train/fully_async_trainer.py#L314C1-L315C1

**Comment by erictang000** [comment]
@SachinKonan fixed this in #873, thanks for asking and helping clear this up!
